# Quality assessment

We are going to begin our single-cell analysis by loading in the output from CellRanger. We will load in our different samples, create a Seurat object with them, and take a look at the quality of the cells.

The general steps to preprocessing your single-cell data with Seurat:

1. **Create a Seurat object**
2. **Filter low-quality cells**
3. **Merge samples**
4. **Normalize counts**
5. **Find variable features**
6. **Scale data**
7. **Determine PCs for Clustering**
8. **Clustering** - FindNeighbors, FindClusters, RunUMAP

### Setup
Begin by initiating a fresh R script and loading the essential packages. Among these, Seurat takes the spotlight. Renowned for its distinctive data structure and robust tools, Seurat facilitates quality control, analysis, and in-depth exploration of single-cell RNA sequencing data. Widely embraced, it stands as a staple in the realm of single-cell RNA analysis, setting a benchmark for its comprehensive functionality and reliability.

```R
library("Seurat") 
library("ggplot2") # creating plots
library("cowplot") # add-on to ggplot, we use the plot_grid function to put multiple violin plots in one image file
library("dplyr")   # a set of functions for data frame manipulation -- a core package of Tidyverse (THE R data manipulation package)
library("Matrix")  # a set of functions for operating on matrices
library("viridis") # color maps for graphs that are more readable than default colors
library("gprofiler2")

datadir = "/project/biocompworkshop/Data_Vault/scRNAseq"
outdir = "/project/biocompworkshop/rshukla/scRNASeq_Results"

setwd(datadir)
list.files()
```

### Initiating Data Import for a Single Sample
Next, we'll import our data matrix generated by Cellranger. Among Cellranger's output files, we'll focus on `sample_filtered_feature_bc_matrix.h5`, containing raw read counts for all cells per sample. Although not human-readable, you can envision it as a table where cell barcodes are column headers and genes (or features) are row names.

We're working with 3 replicates, grouped into pairs of ICB-treated and CD4-depleted ICB-treated samples.

The `CreateSeuratObject` function takes 4 parameters. First is the counts, sourced from the h5 files. The `project` parameter defines the Seurat object's identification, set as the sample name. We also add some filtering criteria: each gene must appear in at least 10 cells, and each cell must have at least 100 features (genes).

For example, processing two samples might look like this:

```R
# Read in the cellranger h5 file
Rep1_ICBdT_data = Read10X_h5("Rep1_ICBdT-sample_filtered_feature_bc_matrix.h5") 

Rep1_ICB_data = Read10X_h5("Rep1_ICB-sample_filtered_feature_bc_matrix.h5")

Rep1_ICBdT_data_seurat_obj = CreateSeuratObject(counts = Rep1_ICBdT_data, project = "Rep1_ICBdT", min.cells = 10, min.features = 100)

Rep1_ICB_data_seurat_obj = CreateSeuratObject(counts = Rep1_ICB_data, project = "Rep1_ICB", min.cells = 10, min.features = 100)
```
### Exploring the Seurat Object Structure
Let's familiarize ourselves with the structure of this Seurat object and understand how we can extract the information necessary for quality assessment, clustering, and subsequent analysis.

```R
print(Rep1_ICB_data_seurat_obj)
```

```R
An object of class Seurat 
14423 features across 4169 samples within 1 assay 
Active assay: RNA (14423 features, 0 variable features)
 1 layer present: counts
```
We observe that our Seurat object comprises 14,423 features representing genes, 4,169 samples representing cells, and 1 assay representing our counts matrix.

```R
Assays(Rep1_ICB_data_seurat_obj) # seurat objects are made up of assays

Layers(Rep1_ICB_data_seurat_obj) # assays contain layers
```



