# Differential expression analysis

In this section, we'll utilize the previously generated Seurat object that has undergone various preprocessing steps, clustering, and cell typing. We'll use this object for gene expression and differential expression analyses. Two key differences in the differential expression analyses conducted in this module are:

- **Identification of Epithelial Cells:** We'll start by identifying epithelial cells in our data using Epcam expression as a marker, considering that tumor cells should exhibit epithelial characteristics. Subsequently, we'll perform a differential expression analysis within the Epcam-positive population(s).

- **Comparison of T Cell Phenotypes:** Another analysis will compare T cell populations in mice treated with ICB therapy against those with T cells depleted, which underwent ICB therapy. This comparison aims to identify differences in T-cell phenotypes related to the treatment.

### Setup

If the saved Seurat object from the previous step isn't already loaded in your current R session, read it into the session.

```R
library(Seurat)
library(dplyr)
library(EnhancedVolcano)
library(presto)

datadir =
setwd(datadir)
merged <- readRDS('preprocessed_object.rds')
```

### Analysis of gene expression in epithelial cells.

Utilize Seurat’s FeaturePlot function to assign colors to each cell based on its Epcam expression on a UMAP plot. FeaturePlot necessitates a minimum of two arguments: the Seurat object and the specific 'feature' to be plotted (such as a gene, PC scores, or any metadata column). For further customization of the FeaturePlot, consult Seurat's documentation [here](link to documentation).

```R
FeaturePlot(merged, features = 'Epcam')
```

Although there are scattered Epcam-positive cells across the UMAP plot, two distinct cell groups are noticeable, potentially representing malignant cells.

Identifying these clusters can be approached in several ways. One method is to use the DimPlot plotting function in conjunction with the FeaturePlot function, separating plots using the + symbol to display multiple plots side-by-side.

```R
DimPlot(merged, group.by = 'seurat_clusters_res0.8', label = TRUE) + 
FeaturePlot(merged, features = 'Epcam') + 
DimPlot(merged, group.by = 'immgen_singler_main', label = TRUE)
```

The plots generated by the preceding commands highlight clusters 9 and 12 as the clusters of interest, but at times, it can be challenging to discern the exact cluster from the UMAP plot alone due to potential overlap between clusters. In such instances, a violin plot (VlnPlot) can offer more clarity. Like FeaturePlot, VlnPlot also utilizes the Seurat object and features as inputs. Additionally, it requires a group.by argument to define the x-axis groupings of the cells. For detailed customization of a Violin plot, please consult the Seurat documentation.

```R
VlnPlot(merged, group.by = 'seurat_clusters_res0.8', features = 'Epcam')
```

Wonderful! It appears that clusters 9 and 12 exhibit the highest Epcam expression. It's intriguing that they are divided into two distinct clusters. This presents an excellent opportunity to employ differential expression analysis to uncover the differences between these clusters.

### Differential gene expression analysis in epithelial cells.

We can start by narrowing down the Seurat object to focus on the desired cells. This can be achieved using Seurat's subset function, which allows us to create a filtered object based on specific values in the metadata column. However, before applying the subset function, we need to designate the default identity of the Seurat object to the relevant metadata column using the SetIdent function. Once the object is subsetted, we can compare the original and subsetted objects side by side to verify the filtering process. Additionally, we can count the number of cells in each type to ensure consistency.

```R
# Set ident to seurat clusters metadata column and subset object to Epcam positive clusters
merged <- SetIdent(merged, value = 'seurat_clusters_res0.8')
merged_epithelial <- subset(merged, idents = c('9', '12'))

# Confirm that we have subset the object as expected visually using a UMAP
DimPlot(merged, group.by = 'seurat_clusters_res0.8', label = TRUE) + 
DimPlot(merged_epithelial, group.by = 'seurat_clusters_res0.8', label = TRUE)

# Confirm that we have subset the object as expected by looking at the individual cell counts
table(merged$seurat_clusters_res0.8)
table(merged_epithelial$seurat_clusters_res0.8)
```

Next, we'll utilize Seurat’s FindMarkers function to conduct a differential expression analysis between the two groups. To use FindMarkers effectively, we also need to use SetIdent to adjust the default ‘Ident’ to the metadata column we want for our comparison. For additional details on FindMarkers, refer to this resource.

Please note that in this analysis, we are using FindMarkers to compare clusters 9 and 12. The standard syntax of FindMarkers requires that we designate each group of cells as ident.1 and ident.2. The output of FindMarkers is a table containing genes that are differentially expressed along with their corresponding log2FC values. The log2FC direction is based on ident.1 in relation to ident.2; genes upregulated in ident.1 have a positive log2FC, while those downregulated have a negative log2FC. 

We include the min.pct=0.25 argument to test genes expressed in at least 25% of cells in either ident.1 or ident.2 groups, reducing false positives compared to the default value of 1%. Additionally, the logfc.threshold=0.1 parameter ensures our results only include genes with a fold change greater than 0.1 or less than -0.1. Adjusting these parameters can improve function efficiency by reducing the number of genes tested.

```R
# Carry out DE analysis between both groups
merged_epithelial <- SetIdent(merged_epithelial, value = "seurat_clusters_res0.8")
epithelial_de <- FindMarkers(merged_epithelial, ident.1 = "9", ident.2 = "12", min.pct=0.25, logfc.threshold=0.1) #how cluster 9 changes wrt cluster 12
```

Upon opening the dataframe "epithelial_de" in your RStudio session, you'll notice it contains genes listed as row names, along with the following columns: p_val, avg_log2FC, pct.1, pct.2, and p_val_adj. The p-values are determined by the test applied during the FindMarkers analysis, while the adjusted p-value reflects the Bonferroni correction test. The columns pct.1 and pct.2 represent the percentage of cells where each gene is detected in the ident.1 and ident.2 groups, respectively.

We can proceed by filtering this dataframe to include only differentially expressed (DE) genes with a significant p-value. Then, we can further filter the resulting dataframe to include only the top 20 genes with the highest absolute log2FC among the significant DE genes. Considering the absolute log2FC enables us to identify both upregulated and downregulated genes.

```R
# Restrict differentially expressed genes to those with an adjusted p-value less than 0.001 
epithelial_de_sig <- epithelial_de[epithelial_de$p_val_adj < 0.001,] 

# Get the top 20 genes by fold change
epithelial_de_sig %>%
  top_n(n = 20, wt = abs(avg_log2FC)) -> epithelial_de_sig_top20
```

The dataframe "epithelial_de_sig_top20" contains the top 20 genes with the most significant differential expression based on log2FC.

We have multiple options for visualizing the differentially expressed genes. We will begin with the Violin and Feature plots as previously discussed. Additionally, we can use a DotPlot to display the differential expression, showing both the average expression of a gene and the percentage of cells expressing it. Apart from these Seurat functions, we can create a volcano plot using the EnhancedVolcano package. In the volcano plot, we can utilize the unfiltered DE results to color and label genes based on specified parameters (pCutoff, FCcutoff).









