# Differential expression analysis

In this section, we'll utilize the previously generated Seurat object that has undergone various preprocessing steps, clustering, and cell typing. We'll use this object for gene expression and differential expression analyses. Two key differences in the differential expression analyses conducted in this module are:

- **Epithelial Cells Comparison:** We'll start by identifying epithelial cells in our data using Epcam expression as a marker, considering that tumor cells should exhibit epithelial characteristics. Subsequently, we'll perform a differential expression analysis within the Epcam-positive population(s).

- **Cell-Type-Specific Phenotype Comparison:** Another analysis will compare T cell populations in mice treated with ICB therapy against those with T cells depleted, which underwent ICB therapy. This comparison aims to identify differences in T-cell phenotypes related to the treatment.

### Setup

If the saved Seurat object from the previous step isn't already loaded in your current R session, read it into the session.

```R
library(Seurat)
library(dplyr)
library(EnhancedVolcano)
library(presto)

datadir = "/project/biocompworkshop/rshukla/scRNASeq_Results"
setwd(datadir)
merged <- readRDS('preprocessed_object.rds')
```

### Analysis of gene expression in epithelial cells.

Utilize Seurat’s FeaturePlot function to assign colors to each cell based on its Epcam expression on a UMAP plot. FeaturePlot necessitates a minimum of two arguments: the Seurat object and the specific 'feature' to be plotted (such as a gene, PC scores, or any metadata column). For further customization of the FeaturePlot, consult Seurat's documentation [here](link to documentation).

```R
FeaturePlot(merged, features = 'Epcam')
```

Although there are scattered Epcam-positive cells across the UMAP plot, two distinct cell groups are noticeable, potentially representing malignant cells.

Identifying these clusters can be approached in several ways. One method is to use the DimPlot plotting function in conjunction with the FeaturePlot function, separating plots using the + symbol to display multiple plots side-by-side.

```R
DimPlot(merged, group.by = 'seurat_clusters_res0.8', label = TRUE) + 
FeaturePlot(merged, features = 'Epcam') + 
DimPlot(merged, group.by = 'immgen_singler_main', label = TRUE)
```

The plots generated by the preceding commands highlight clusters 9 and 12 as the clusters of interest, but at times, it can be challenging to discern the exact cluster from the UMAP plot alone due to potential overlap between clusters. In such instances, a violin plot (VlnPlot) can offer more clarity. Like FeaturePlot, VlnPlot also utilizes the Seurat object and features as inputs. Additionally, it requires a group.by argument to define the x-axis groupings of the cells. For detailed customization of a Violin plot, please consult the Seurat documentation.

```R
VlnPlot(merged, group.by = 'seurat_clusters_res0.8', features = 'Epcam')
```

Wonderful! It appears that clusters 9 and 12 exhibit the highest Epcam expression. It's intriguing that they are divided into two distinct clusters. This presents an excellent opportunity to employ differential expression analysis to uncover the differences between these clusters.

### Differential gene expression analysis in epithelial cells.

We can start by narrowing down the Seurat object to focus on the desired cells. This can be achieved using Seurat's subset function, which allows us to create a filtered object based on specific values in the metadata column. However, before applying the subset function, we need to designate the default identity of the Seurat object to the relevant metadata column using the SetIdent function. Once the object is subsetted, we can compare the original and subsetted objects side by side to verify the filtering process. Additionally, we can count the number of cells in each type to ensure consistency.

```R
# Set ident to seurat clusters metadata column and subset object to Epcam positive clusters
merged <- SetIdent(merged, value = 'seurat_clusters_res0.8')
merged_epithelial <- subset(merged, idents = c('9', '12'))

# Confirm that we have subset the object as expected visually using a UMAP
DimPlot(merged, group.by = 'seurat_clusters_res0.8', label = TRUE) + 
DimPlot(merged_epithelial, group.by = 'seurat_clusters_res0.8', label = TRUE)

# Confirm that we have subset the object as expected by looking at the individual cell counts
table(merged$seurat_clusters_res0.8)
table(merged_epithelial$seurat_clusters_res0.8)
```

Next, we'll utilize Seurat’s `FindMarkers` function to conduct a differential expression analysis between the two groups. To use `FindMarkers` effectively, we also need to use `SetIdent` to adjust the default ‘Ident’ to the metadata column we want for our comparison. 

Please note that in this analysis, we are using `FindMarkers` to compare clusters `9` and `12`. The standard syntax of `FindMarkers` requires that we designate each group of cells as `ident.1` and `ident.2`. The output of `FindMarkers` is a table containing genes that are differentially expressed along with their corresponding `log2FC` values. The `log2FC` direction is based on `ident.1` in relation to `ident.2`; genes upregulated in `ident.1` have a positive `log2FC`, while those downregulated have a negative `log2FC`. 

We include the `min.pct=0.25` argument to test genes expressed in at least 25% of cells in either `ident.1` or `ident.2` groups, reducing false positives compared to the default value of 1%. Additionally, the `logfc.threshold=0.1` parameter ensures our results only include genes with a fold change greater than 0.1 or less than -0.1. Adjusting these parameters can improve function efficiency by reducing the number of genes tested.

```R
# Carry out DE analysis between both groups
merged_epithelial <- SetIdent(merged_epithelial, value = "seurat_clusters_res0.8")
epithelial_de <- FindMarkers(merged_epithelial, ident.1 = "9", ident.2 = "12", min.pct=0.25, logfc.threshold=0.1) #how cluster 9 changes wrt cluster 12
```

Upon opening the dataframe "epithelial_de" in your RStudio session, you'll notice it contains genes listed as row names, along with the following columns: `p_val`, `avg_log2FC`, `pct.1`, `pct.2`, and `p_val_adj`. The p-values are determined by the test applied during the `FindMarkers` analysis, while the adjusted p-value reflects the Bonferroni correction test. The columns `pct.1` and `pct.2` represent the percentage of cells where each gene is detected in the `ident.1` and `ident.2` groups, respectively.

We can proceed by filtering this dataframe to include only differentially expressed (DE) genes with a significant p-value. Then, we can further filter the resulting dataframe to include only the top 20 genes with the highest absolute `log2FC` among the significant DE genes. Considering the absolute `log2FC` enables us to identify both upregulated and downregulated genes.

```R
# Restrict differentially expressed genes to those with an adjusted p-value less than 0.001 
epithelial_de_sig <- epithelial_de[epithelial_de$p_val_adj < 0.001,] 

# Get the top 20 genes by fold change
epithelial_de_sig %>%
  top_n(n = 20, wt = abs(avg_log2FC)) -> epithelial_de_sig_top20
```

The dataframe "epithelial_de_sig_top20" contains the top 20 genes with the most significant differential expression based on `log2FC`.

We have multiple options for visualizing the differentially expressed genes. We will begin with the Violin and Feature plots as previously discussed. Additionally, we can use a DotPlot to display the differential expression, showing both the average expression of a gene and the percentage of cells expressing it. Apart from these Seurat functions, we can create a volcano plot using the `EnhancedVolcano` package. In the volcano plot, we can utilize the unfiltered DE results to color and label genes based on specified parameters (`pCutoff`, `FCcutoff`).

```R
# Get list of top 20 DE genes for ease
epithelial_de_sig_top20_genes <- rownames(epithelial_de_sig_top20)

# Plot all 20 genes in violin plots
VlnPlot(merged_epithelial, features = epithelial_de_sig_top20_genes, 
  group.by = 'seurat_clusters_res0.8', ncol = 5, pt.size = 0)

# Plot all 20 genes in UMAP plots
FeaturePlot(merged_epithelial, features = epithelial_de_sig_top20_genes, ncol = 5)

# Plot all 20 genes in a DotPlot
DotPlot(merged_epithelial, features = epithelial_de_sig_top20_genes, 
  group.by = 'seurat_clusters_res0.8') + RotatedAxis()

# Plot all differentially expressed genes in a volcano plot
EnhancedVolcano(epithelial_de,
  lab = rownames(epithelial_de),
  x = 'avg_log2FC',
  y = 'p_val_adj',
  title = 'Cluster9 wrt Cluster 12',
  pCutoff = 0.05,
  FCcutoff = 0.5,
  pointSize = 3.0,
  labSize = 5.0,
  colAlpha = 0.3)
```

Stay tuned to discover how we can interpret the significance of these genes! The upcoming pathway analysis module will provide insights into this. Meanwhile, let's generate a TSV file containing our differential expression (DE) results for future use. We'll need to rerun FindMarkers with slightly adjusted parameters; specifically, we'll set the logfc.threshold parameter to 0. This change is necessary as one of the pathway analysis tools mandates the inclusion of all genes in the analysis (more details on this will follow).

```R
# Rerun FindMarkers
epithelial_de_gsea <- FindMarkers(merged_epithelial, ident.1 = "9", 
  ident.2 = "12", min.pct=0.25, logfc.threshold=0)

# Save this table as a TSV file
write.table(x = epithelial_de_gsea, 
  file = 'epithelial_de_gsea.tsv', sep='\t')
```

### Differential gene expression analysis in T cells.

For the analysis focusing on T cells, our goal is to compare T cells from mice treated with ICB to T cells from mice with some of their T cells depleted and treated with ICB (ICBdT). Initially, we will narrow down our merged object to exclusively include T cells. This involves consolidating various T cell annotations from the cell typing section. We'll begin by identifying all possible cell types available and selecting those related to T cells. Afterward, we'll use `SetIdent` to designate the cell type metadata column and subset the data to include only the cell types corresponding to T cells. Lastly, we'll verify that the subsetting process was successful as expected.

```R
# Check all the annotated celltypes
unique(merged$immgen_singler_main)

# Pick the ones that are related to T cells
t_celltypes_names <- c('T cells', 'NKT', 'Tgd')
merged <- SetIdent(merged, value = 'immgen_singler_main')
merged_tcells <- subset(merged, idents = t_celltypes_names)

# Confirm that we have subset the object as expected visually using a UMAP
DimPlot(merged, group.by = 'immgen_singler_main', label = TRUE) + 
DimPlot(merged_tcells, group.by = 'immgen_singler_main', label = TRUE)

# Confirm that we have subset the object as expected by looking at the individual cell counts
table(merged$immgen_singler_main)
table(merged_tcells$immgen_singler_main)
```

Let's proceed with comparing T cells from mice treated with ICB versus those treated with ICBdT. Initially, we must differentiate between the ICB and ICBdT cells. Begin by selecting the object in RStudio and expanding the `meta.data` section to view the columns and the type of data they contain.

It appears that the `orig.ident` column contains information about both the condition and replicates. However, for this differential expression (DE) analysis, we require a `meta.data` column that consolidates the replicates for each condition. Therefore, our goal is to merge the replicates of each condition into a unified category.

```R
# We'll start by checking the possible names each replicate has.
unique(merged_tcells$orig.ident)

# There are 6 possible values, 3 replicates for the ICB treatment condition, and 3 for the ICBdT condition
# so we can combine "Rep1_ICB", "Rep3_ICB", "Rep5_ICB" to ICB, and "Rep1_ICBdT", "Rep3_ICBdT", "Rep5_ICBdT" to ICBdT. 
# first initialize a metadata column for experimental_condition
merged_tcells@meta.data$experimental_condition <- NA

# Now we can take all cells that are in each replicate-condition, 
# and assign them to the appropriate condition
merged_tcells@meta.data$experimental_condition[merged_tcells@meta.data$orig.ident %in% c("Rep1_ICB", "Rep3_ICB", "Rep5_ICB")] <- "ICB"
merged_tcells@meta.data$experimental_condition[merged_tcells@meta.data$orig.ident %in% c("Rep1_ICBdT", "Rep3_ICBdT", "Rep5_ICBdT")] <- "ICBdT"

# Double check that the new column we generated makes sense 
# (each replicate should correspond to its experimental condition)
table(merged_tcells@meta.data$orig.ident, merged_tcells@meta.data$experimental_condition)
```

Now that we have established the experimental conditions, we can proceed to compare the T cells from both groups. Initially, we will utilize `FindMarkers` with parameters similar to the previous analysis to observe the changes in ICBdT compared to ICB. Following this, we will filter the dataframe to include only significant genes and examine the top five genes that are most upregulated and downregulated based on log2FC.

```R
# Carry out DE analysis between both groups
merged_tcells <- SetIdent(merged_tcells, value = "experimental_condition")
tcells_de <- FindMarkers(merged_tcells, ident.1 = "ICBdT", ident.2 = "ICB", min.pct=0.25)

# Restrict differentially expressed genes to those with an adjusted p-value less than 0.001 
tcells_de_sig <- tcells_de[tcells_de$p_val_adj < 0.001,]

# Find the top 5 most downregulated genes
tcells_de_sig %>%
  top_n(n = 5, wt = -avg_log2FC)
# Find the top 5 most upregulated genes
tcells_de_sig %>%
  top_n(n = 5, wt = avg_log2FC)
```

The gene with the highest downregulation in the ICBdT condition in terms of fold change is `Cd4`. This finding is logical since the monoclonal antibody (GK1.5) used for T cell depletion specifically targets CD4 T cells. 

It's noteworthy that among the list of upregulated genes, `Cd8b1` appears. It might be intriguing to investigate whether the `CD8` T cells' phenotype alters depending on the treatment condition. Let's proceed by narrowing down the object to include only `CD8` T cells, identifying differentially expressed (DE) genes to observe how ICBdT `CD8` T cells differ from ICB `CD8` T cells, and visualize these results similar to our previous approach.

```R
# Subset object to CD8 T cells. Since we already showed how to subset cells using the clusters earlier, 
# this time we'll subset to CD8 T cells by selecting for cells with high 
# expression of Cd8 genes and low expression of Cd4 genes
merged_cd8tcells <- subset(merged_tcells, subset= Cd8b1 > 1 & Cd8a > 1 & Cd4 < 0.1)

# Carry out DE analysis between both groups
merged_cd8tcells <- SetIdent(merged_cd8tcells, value = "experimental_condition")
cd8tcells_de <- FindMarkers(merged_cd8tcells, ident.1 = "ICBdT", ident.2 = "ICB", min.pct=0.25) #how ICBdT changes wrt ICB

# Restrict differentially expressed genes to those with an adjusted p-value less than 0.001 
cd8tcells_de_sig <- cd8tcells_de[cd8tcells_de$p_val_adj < 0.001,]

# Get the top 20 genes by fold change
cd8tcells_de_sig %>%
  top_n(n = 20, wt = abs(avg_log2FC)) -> cd8tcells_de_sig_top20

# Get list of top 20 DE genes for ease
cd8tcells_de_sig_top20_genes <- rownames(cd8tcells_de_sig_top20)


# Plot all 20 genes in violin plots
VlnPlot(merged_cd8tcells, features = cd8tcells_de_sig_top20_genes, group.by = 'experimental_condition', ncol = 5, pt.size = 0)

# Plot all 20 genes in UMAP plots
FeaturePlot(merged_cd8tcells, features = cd8tcells_de_sig_top20_genes, ncol = 5)

# Plot all 20 genes in a DotPlot
DotPlot(merged_cd8tcells, features = cd8tcells_de_sig_top20_genes, group.by = 'experimental_condition') + RotatedAxis()

#plot all differentially expressed genes in a volcano plot
EnhancedVolcano(cd8tcells_de,
  lab = rownames(cd8tcells_de),
  x = 'avg_log2FC',
  y = 'p_val_adj',
  title = 'ICBdT wrt ICB',
  pCutoff = 0.05,
  FCcutoff = 0.5,
  pointSize = 3.0,
  labSize = 5.0,
  colAlpha = 0.3)
```

Now, you can begin conducting literature searches on some of these genes. You may discover that antigen-specific memory `CD8` T cells exhibit increased expression of `Bcl2` and `Cdk8`.

Alternatively, we can utilize gene set and pathway analyses to explore the potential cellular processes in which these cells may be involved.





